builder/slide_and_fold_storage.cc

```diff
@@ -239,39 +364,28 @@ class slide_and_fold : public node_mutator {
-        if (!bounds) continue;
 
         // We don't want to use the bounds of the loop we are sliding over here.
         auto ignore_loop_bounds = set_value_in_scope(*loop.expr_bounds, loop.sym, interval_expr::all());
 
         expr loop_var = variable::make(loop.sym);
-
-        for (int d = 0; d < static_cast<int>(bounds->size()); ++d) {
-          interval_expr cur_bounds_d = (*bounds)[d];
-          if (!depends_on(cur_bounds_d, loop.sym).any()) {
-            // TODO: In this case, the func is entirely computed redundantly on every iteration. We should be able to
-            // just compute it once.
-            continue;
-          }
-
-          interval_expr prev_bounds_d = {
-              substitute(cur_bounds_d.min, loop.sym, loop_var - loop.step),
-              substitute(cur_bounds_d.max, loop.sym, loop_var - loop.step),
-          };
-
-          // A few things here struggle to simplify when there is a min(loop_max, x) expression involved, where x is
-          // some expression that is bounded by the loop bounds. This min simplifies away if we know that x <= loop_max,
-          // but the simplifier can't figure that out. As a hopefully temporary workaround, we can just substitute
-          // infinity for the loop max.
-          auto ignore_loop_max = [&](const expr& e) { return substitute(e, loop.bounds.max, positive_infinity()); };
-
-          interval_expr overlap = prev_bounds_d & cur_bounds_d;
-          if (prove_true(ignore_loop_max(overlap.empty()))) {
-            // The bounds of each loop iteration do not overlap. We can't re-use work between loop iterations, but we
-            // can fold the storage.
-            expr fold_factor = simplify(bounds_of(ignore_loop_max(cur_bounds_d.extent()), *loop.expr_bounds).max);
-            if (is_finite(fold_factor) && !depends_on(fold_factor, loop.sym).any()) {
+        auto ff = fold_factors[output];
+        if (!ff) continue;
+        for (int d = 0; d < static_cast<int>(ff->size()); ++d) {
+          expr fold_factor = (*ff)[d].first;
+          expr overlap = (*ff)[d].second;
+          if (!is_finite(fold_factor)) continue;
+
+          if (!depends_on(fold_factor, loop.sym).any()) {
+            if (overlap.defined()) {
+              // We need an extra fold per worker when parallelizing the loop.
+              // TODO: This extra folding seems excessive, it allows all workers to execute any stage.
+              // If we can figure out how to add some synchronization to limit the number of workers that
+              // work on a single stage at a time, we should be able to reduce this extra folding.
+              fold_factor += (loop.worker_count - 1) * overlap;
+              // Align the fold factor to the loop step size, so it doesn't try to crop across a folding boundary.
+              fold_factor = simplify(align_up(fold_factor, loop.step));
+            } else {
               if (loop.add_synchronization()) {
```

Comment:
Related to my other comment above, hopefully this just isn't needed, we already add synchronziation above, this is just checking if we want it, but I don't think we need to do that.